src/PlonkVerify.hs

```diff
@@ -1,27 +1,172 @@
--- This file was generated by `hs-bindgen` crate and contains C FFI bindings
--- wrappers for every Rust function annotated with `#[hs_bindgen]`
-
+{-# LANGUAGE DeriveLift #-}
+{-# LANGUAGE DerivingStrategies #-}
 {-# LANGUAGE ForeignFunctionInterface #-}
+{-# LANGUAGE ImportQualifiedPost #-}
+{-# LANGUAGE TemplateHaskell #-}
 
--- Why not rather using `{-# LANGUAGE CApiFFI #-}` language extension?
---
--- * Because it's GHC specific and not part of the Haskell standard:
---   https://ghc.gitlab.haskell.org/ghc/doc/users_guide/exts/ffi.html ;
---
--- * Because the capabilities it gave (by rather works on top of symbols of a C
---   header file) can't work in our case. Maybe we want a future with an
---   {-# LANGUAGE RustApiFFI #-} language extension that would enable us to
---   work on top of a `.rs` source file (or a `.rlib`, but this is unlikely as
---   this format has purposely no public specification).
+-- This library is not tight to a particular verifying key. But it provides
+-- tools that help users to manage verifying keys in their applications.
+
+module PlonkVerify
+( verify_plonk_bn254
+, VKey(..)
+, Proof(..)
+, ProgramId(..)
+, PublicParameter(..)
+, PublicParameterHash(..)
+, mkPublicParameterHash
+, hashPublicParameters
+, verifyPlonkBn254
+, verifyPlonkBn254'
+) where
 
-{-# OPTIONS_GHC -Wno-unused-imports #-}
+import Crypto.Hash.SHA256 qualified as H
 
-module PlonkVerify (verify_plonk_bn254) where
+import Data.ByteString qualified as B
+import Data.ByteString.Base16 qualified as B16
+import Data.ByteString.Short qualified as BS
+import Data.Foldable
 
-import Data.Int
-import Data.Word
 import Foreign.C.String
 import Foreign.C.Types
-import Foreign.Ptr
 
-foreign import ccall safe "__c_verify_plonk_bn254" verify_plonk_bn254 :: CString -> CString -> CString -> CString -> IO (CUInt)
\ No newline at end of file
+import Language.Haskell.TH.Syntax (Lift)
+
+import System.IO.Temp
+import Data.Bits
+
+-- -------------------------------------------------------------------------- --
+-- Plonk Verifier FFI
+
+foreign import ccall safe "__c_verify_plonk_bn254"
+    verify_plonk_bn254 :: CString -> CString -> CString -> CString -> IO (CUInt)
+
+-- -------------------------------------------------------------------------- --
+-- API Types
+
+-- | The verifying key identifies the circuit of the VM that is used to execute
+-- and proof the program. Intuitively, one can think of the key as a version
+-- identifier of the VM.
+--
+-- When the VM circuit is updated the key changes. Applications should record
+-- what key a proofs depends on for verification. The module "EmbedVKey"
+-- provides tools for embedding existing keys from files into applications
+-- binaries.
+--
+newtype VKey = VKey
```

Comment:
Maybe `MachineVKey` would be a better name for this type?

Reply:
`VMKey` is probably the compromise name.